enclave/crypto/openssl/cert.c

```diff
@@ -0,0 +1,457 @@
+ */
+char* decode_oid_to_str(char* oid, int oid_size)
+{
+    char* oid_str = NULL;
+    char tmp_buf[40];
+    char oid_buf[OE_OID_MAX_LENGTH] = {0};
+    char byte = 0;
+    long block = 0;
+    int index = 0;
+    int len = 0;
+
+    if (!oid || oid_size <= 0)
+        goto done;
+
+    /*
+     * Break the hex string into blocks, each of which represents a integer,
+     * and then concatenate these integers into a dot-notation string.
+     */
+    while (index < oid_size)
+    {
+        block = 0;
+        byte = oid[index++];
+        /*
+         * A block ends with a byte with bit 8 cleared. If the bit 8 is set,
+         * appending the lower 7 bits of the byte into the current block.
+         */
+        while (byte & 0x80)
+        {
+            block <<= 7L;
+            block |= byte & 0x7f;
+            /*
+             * Limit the value of a block to 8 bytes (align with OpenSSL
+             * implementation). See ASN1_get_object in asn1_lib.c for the
+             * reference.
+             */
+            if (block > (OE_INT_MAX >> 7L))
+                goto done;
+
+            byte = oid[index++];
+            if (index >= oid_size)
+                goto done;
+        }
+        /* Append the ending byte to the block. */
+        block <<= 7L;
+        block |= byte & 0x7f;
+
+        /*
+         * Zero length of oid_buf indicates that we're decoding the first block,
+         * which requires special handling. The value of the block (z) is
+         * decoded into two integers (x and y). The rule is as follows. z = 40 *
+         * x + y, where x = 0, 1, or 2.
+         */
+        if (strlen(oid_buf) == 0)
+        {
+            long first;
+            long second;
+
+            first = block / 40;
+            if (first > 2)
+                first = 2;
+
+            second = block - 40 * first;
+            len = snprintf(tmp_buf, sizeof(tmp_buf), "%ld.%ld", first, second);
+        }
+        else
+            len = snprintf(tmp_buf, sizeof(tmp_buf), ".%ld", block);
+
+        if (len <= 0)
+            goto done;
+        if (strlen(oid_buf) + (size_t)len >= OE_OID_MAX_LENGTH)
+            goto done;
+
+        strncat(oid_buf, tmp_buf, (size_t)len);
+    }
+
+    /* Copy from buffer */
+    size_t oid_len = strlen(oid_buf) + 1;
+    oid_str = (char*)oe_malloc(oid_len);
+    if (oid_str == NULL)
+        goto done;
+
+    strncpy(oid_str, oid_buf, oid_len);
+
+done:
+    return oid_str;
+}
+
+oe_result_t oe_gen_custom_x509_cert(
+    oe_cert_config_t* config,
+    unsigned char* cert_buf,
+    size_t cert_buf_size,
+    size_t* bytes_written)
+{
+    oe_result_t result = OE_CRYPTO_ERROR;
+    X509* x509cert = NULL;
+    X509V3_CTX ctx;
+    BIO* bio = NULL;
+    X509_NAME* name = NULL;
+    EVP_PKEY* subject_issuer_key_pair = NULL;
+    X509_EXTENSION* ext = NULL;
+    ASN1_OBJECT* obj = NULL;
+    ASN1_OCTET_STRING* data = NULL;
+    BASIC_CONSTRAINTS* bc = NULL;
+    unsigned char* buff = NULL;
```

Comment:
Style: prefer verbose names over abbreviations (this is inconsistently abbreviated as `buf` elsewhere in the codebase for example)

Reply:
Updated.